Connect-DbaInstance / new code path - General improvements

andreasjordan opened the pull request:
## Type of Change
 - [ ] Bug fix (non-breaking change, fixes #<!--issue number--> )
 - [x] New feature (non-breaking change, adds functionality, fixes #7425, fixes #7426 )
 - [ ] Breaking change (effects multiple commands or functionality, fixes #<!--issue number--> )
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] If new file reference added for test, has is been added to github.com/sqlcollaborative/appveyor-lab ?
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system
 
### Purpose
All necessary changes to later use the new code path as the default.

### Approach
I addressed most of the TODO entries and added some new features, so beeing able to use more parameters together with a server SMO as SqlInstance.

I would like to discuss the different topics in the linked issues to have a better overview.
If the majority of the maintainers rejects the changes related to the linked issues, I will of course remove them.

potatoqualitee commented:
thanks much. Is this the same code that was in experimental for a while? if so, i can merge it then hold off on publishing dev for a few days

andreasjordan commented:
No, new functionality has also been added. I'll walk through the diff, naming the new lines.
* 476 - 537: Only documentation
* 581 - 632: New functionality, see #7425
* 639 - 658: Additional tests for parameters that are not used anymore by the new code path.
* 661 - 662: No support for MFA inside of the command, possible with outside usage of the AZ module and AccessToken
* 698 - 899: Changed or added some debug messages
* 907 - 919: I changed the way the connection is opened. After many tests this seems to be we best way.
* 940 - 972: New functionality, but also changing the default behavior of the new code path to align it with the old path. But for those how have used the new code path there maybe situations where they get a different value for ComputerName. But with the new config parameter now every desired value can be achieved. See #7426 
* 974 - 1007: New functionality, see #7427 
* 1028 - 1038: Just copied from the old code path and added some documentation
* 1040 - 1075: Copied from the old code path but hiding it behind a configuration. I can't see an effect of this, have traced the sql server and compared what calls are send to the server on connect - I don't see any difference. Because I don't like code where I don't know that it has a positive effect, but on the other hand I just might not see it - I choose this. As soon as someone can prove and show an example where this has a positive effect - we just enable it by default. If in a one year time, nobody came to us - we delete the code and copy it to an issue on github for backup.

So wrapping up: I would like to have your opinion on #7426 and #7427 and of cource the other functional changes. If we have the code we want, then I would like to have these changes for a month or two in the non-default experimental code path. But I don't have any further changes, so the next step then would be just to change the default and add a config parameter to activate the legacy code path as a backup.

andreasjordan commented:
If you would like to talk about the changes, I can set up a Teams meeting and we can go through the code together.

potatoqualitee commented:
This is incredible, thank you so much. Yes, I'll likely need a teams meeting @andreasjordan 💯  Perhaps late next week? This is intense and I can't think in this heat.

andreasjordan commented:
Just ping me on slack if you have time.

wsmelton commented on functions/Connect-DbaInstance.ps1:
```diff
@@ -462,36 +462,58 @@ function Connect-DbaInstance {
             #>
+
+            # The following line is just added so that we can use AppVayor to run all test with the experimental code path
+            Set-DbatoolsConfig -FullName sql.connection.experimental -Value $true
```
FYI, if you want Appveyor to run using this easier to put it in the Test itself.

andreasjordan replied:
But I want ALL tests, so all commands that run, to use this. I would have to put it in some global Appveyor config?

potatoqualitee replied:
yeah, let me figure this out

potatoqualitee replied:
done! added to \tests\appveyor.pester.ps1

wsmelton commented on functions/Connect-DbaInstance.ps1:
```diff
@@ -665,26 +725,21 @@ function Connect-DbaInstance {
+                    if ($authType -eq 'azure integrated') {
                         # Azure AD integrated security
-                        # TODO: This is not tested
+                        # TODO: This is not tested / How can we test that?
```
It requires a federated Azure AD to test 😉 (which is not simple to build-out).

potatoqualitee commented:
Added a new containers test that performs a migration and sets up mirroring and it worked 🥳  

It performs a migration using the old code path then the new code path!

https://github.com/sqlcollaborative/dbatools/pull/7420/checks?check_run_id=2978360938

potatoqualitee commented:
Something we may have to keep an eye on: https://ci.appveyor.com/project/sqlcollaborative/dbatools/builds/39854724/job/h6bmlcliith86qrm

It'll be a 60 minute timeout. Possibly Appveyor but possibly not.

potatoqualitee commented:
Old tests pass, new tests pass, removed Config, bout to merge

andreasjordan commented:
> Something we may have to keep an eye on: https://ci.appveyor.com/project/sqlcollaborative/dbatools/builds/39854724/job/h6bmlcliith86qrm
> 
> It'll be a 60 minute timeout. Possibly Appveyor but possibly not.

I don't know what's behind the link - "Build not found or access denied.". Maybe add a screenshot?

potatoqualitee commented:
ahh, it was just a timeout on a Copy command. So we can generically keep an eye out for timeouts.
